okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java

```diff
@@ -925,96 +925,98 @@ private ModuleDescriptor findSystemInterface(Tenant tenant, String interfaceName
+            Future<Void> future = Future.succeededFuture();
+            if (options.getAsync()) {
+              List<TenantModuleDescriptor> tml2 = new LinkedList<>();
+              for (TenantModuleDescriptor tm : tml) {
+                tml2.add(tm.cloneWithoutStage());
+              }
+              promise.complete(tml2);
+            }
+            future = future.compose(x -> {
+              for (TenantModuleDescriptor tm : tml) {
+                tm.setStage(TenantModuleDescriptor.Stage.pending);
               }
-              promise1.complete();
+              return jobs.put(t.getId(), job.getId(), job);
             });
-            future = future.compose(x -> promise1.future());
-          } else {
-            future = future.compose(x -> installTenantModule(t, pc, options, modsAvailable, tm));
-          }
-          future = future.compose(x -> {
-            if (tm.getMessage() != null) {
-              return Future.succeededFuture();
+            if (options.getDeploy()) {
+              future = future.compose(x -> autoDeploy(t, job, modsAvailable, tml));
             }
-            tm.setStage(TenantModuleDescriptor.Stage.done);
-            return jobs.put(t.getId(), job.getId(), job);
-          });
-        }
-        // if we are really upgrading permissions do a refresh last
-        for (TenantModuleDescriptor tm : tml) {
-          if (tm.getAction() == Action.enable && tm.getFrom() != null) {
-            ModuleDescriptor mdTo = modsAvailable.get(tm.getId());
-            // mdFrom is null so reloadPermissions is triggered!!!!
-            future = future.compose(x -> reloadPermissions(t, options, null, mdTo, pc));
-          }
-        }
-        if (options.getDeploy()) {
-          future.compose(x -> autoUndeploy(t, job, modsAvailable, tml));
-        }
-        for (TenantModuleDescriptor tm : tml) {
-          future = future.compose(x -> {
-            if (tm.getMessage() != null) {
-              return Future.succeededFuture();
+            for (TenantModuleDescriptor tm : tml) {
+              future = future.compose(x -> {
+                tm.setStage(TenantModuleDescriptor.Stage.invoke);
+                return jobs.put(t.getId(), job.getId(), job);
+              });
+              if (options.getIgnoreErrors()) {
+                Promise<Void> promise1 = Promise.promise();
+                installTenantModule(t, pc, options, modsAvailable, tm).onComplete(x -> {
+                  if (x.failed()) {
+                    logger.warn("Ignoring error for tenant {} module {}",
+                        t.getId(), tm.getId(), x.cause());
+                  }
+                  promise1.complete();
+                });
+                future = future.compose(x -> promise1.future());
+              } else {
+                future = future
+                    .compose(x -> installTenantModule(t, pc, options, modsAvailable,tm));
+              }
```

Comment:
```suggestion
              future = future.compose(x -> installTenantModule(t, pc, options, modsAvailable,tm));
              if (options.getIgnoreErrors()) {
                future = future.otherwise(e -> {
                  logger.warn("Ignoring error for tenant {} module {}", t.getId(), tm.getId(), e);
                  return null;
                });
              }
```

Reply:
oh.. yes. Had a similar change here, but using recover. Using otherwise is probably more elegant.